src/coreclr/jit/lsrabuild.cpp

```diff
@@ -1491,12 +1492,31 @@ void LinearScan::buildUpperVectorSaveRefPositions(GenTree* tree, LsraLocation cu
         assert((fpCalleeKillSet & RBM_FLT_CALLEE_TRASH) != RBM_NONE);
         assert((fpCalleeKillSet & RBM_FLT_CALLEE_SAVED) == RBM_NONE);
 
-        // We only need to save the upper half of any large vector vars that are currently live.
-        VARSET_TP       liveLargeVectors(VarSetOps::Intersection(compiler, currentLiveVars, largeVectorVars));
-        VarSetOps::Iter iter(compiler, liveLargeVectors);
-        unsigned        varIndex = 0;
-        bool            blockAlwaysReturn =
-            compiler->compCurBB->KindIs(BBJ_THROW, BBJ_EHFINALLYRET, BBJ_EHFAULTRET, BBJ_EHFILTERRET, BBJ_EHCATCHRET);
+        // We should only save the upper half of any large vector vars that are currently live.
+        // However, the liveness information may not be accurate, specially around the place where
+        // we load the LCL_VAR and the node that uses it. Hence, as a conservative approach, we will
+        // add all variables that are live-in/defined in the block. We need to add variable although
+        // it is not in the live-out set, because a variable may get defined before the call and
+        // (last) used after the call.
+        //
+        // This will create more UpperSave/UpperRestore RefPositions then needed, but we need to do
+        // this for correctness anyway.
+        VARSET_TP bbLiveDefs(VarSetOps::Union(compiler, compiler->compCurBB->bbLiveIn, compiler->compCurBB->bbVarDef));
+
+        VARSET_TP liveDefsLargeVectors(VarSetOps::Intersection(compiler, bbLiveDefs, largeVectorVars));
+
+        // Make sure that `liveLargeVectors` captures the currentLiveVars as well.
+        VARSET_TP liveLargeVectors(VarSetOps::Intersection(compiler, currentLiveVars, largeVectorVars));
+
+#ifdef DEBUG
+        assert(VarSetOps::IsEmpty(compiler, liveLargeVectors) ||
+               VarSetOps::IsSubset(compiler, liveLargeVectors, liveDefsLargeVectors));
```

Comment:
Nit:
```suggestion
        assert(VarSetOps::IsSubset(compiler, liveLargeVectors, liveDefsLargeVectors));
```
If it's empty it should also be a subset

Reply:
(Can also remove the #ifdef DEBUG)

Reply:
> (Can also remove the #ifdef DEBUG)

yes, I had a code earlier inside this, but i removed it later. So yes, `#ifdef DEBUG` is not needed anymore.

> If it's empty it should also be a subset

true, not sure at one point I was hitting an assert. May be my calculation were wrong. Removed it.